### ISSUE TITLE
Add another abstract variant

### ISSUE THREAD

**Issue by treeowl** [ekmett/bound#46]
```haskell
abstractE :: Monad f => (a -> Either a' b) -> f a -> Scope b f a'
abstractE f e = Scope (liftM k e) where
  k y = case f y of
    Right z  -> B z
    Left q -> F (return q)
```
The idea is that the free variable type can change to reflect the fact that some of its values are gone. I don't know if this is practically important, but a nice side effect is that the type of `abstractE` expresses/constrains what it does much more thoroughly than the type of `abstract` does.

**Comment by phadej** [comment]
This one has a nice consequence of being able to define

```haskell
abstractAll :: Monad f => (a -> b) -> f a -> Scope b f c
```

:+1: from me, @treeowl  can you make a PR?

**Comment by treeowl** [comment]
Closed in #58
